downloadProgressListener_indefinitely worker thread doesn't terminate

As per the discussion [here](https://github.com/realm/realm-java/pull/5159#discussion_r136507434), we need to investigate why the `downloadProgressListener_indefinitely` test fails. Removing the [join](https://github.com/realm/realm-java/pull/5159/files/69aa8f2d39fce182485c34767d6249cd7d8173f3#diff-98189d61fc029709ddad095607a8590eL193) seems to fix the issue, which indicates that the `worker` thread didn't terminate, which causes the test thread to wait indefinitely until it times out

full log:
```
10-16 22:04:43.368 26369-26384/io.realm.test I/TestRunner: started: downloadProgressListener_indefinitely(io.realm.objectserver.ProgressListenerTests)
10-16 22:04:43.369 26369-26369/io.realm.test I/MonitoringInstr: Activities that are still in CREATED to STOPPED: 0
10-16 22:04:43.395 26369-26384/io.realm.test D/REALM_JAVA: Network request (authenticate): http://127.0.0.1:9080/auth
10-16 22:04:43.540 26369-26384/io.realm.test D/REALM_JAVA: AuthenticateResponse. accessToken = null
10-16 22:04:43.540 26369-26384/io.realm.test I/REALM_JAVA: Succeeded authenticating user.
                                                           {UserId: d228000a0019b615b194c5e11027942e, AuthUrl: http://127.0.0.1:9080/auth}
10-16 22:04:43.548 26369-26384/io.realm.test D/REALM_JAVA: first session created add network listener
10-16 22:04:43.551 26369-26384/io.realm.test D/REALM_JNI: Callback to Java requesting token for path
10-16 22:04:43.552 26369-26392/io.realm.test D/REALM_JNI: SyncClient thread created
10-16 22:04:43.553 26369-26393/io.realm.test D/REALM_JAVA: Network request (authenticate): http://127.0.0.1:9080/auth
10-16 22:04:43.565 26369-26384/io.realm.test D/REALM_JNI: Callback to Java requesting token for path
10-16 22:04:43.565 26369-26394/io.realm.test D/REALM_JAVA: Network request (authenticate): http://127.0.0.1:9080/auth
10-16 22:04:43.596 26369-26393/io.realm.test D/REALM_JAVA: AuthenticateResponse. Identity d228000a0019b615b194c5e11027942e; Path /d228000a0019b615b194c5e11027942e/tests
10-16 22:04:43.596 26369-26393/io.realm.test D/REALM_JAVA: Session[/data/data/io.realm.test/cache/junit1946518965/d228000a0019b615b194c5e11027942e/d228000a0019b615b194c5e11027942e/tests]: Access token acquired
10-16 22:04:43.602 26369-26394/io.realm.test D/REALM_JAVA: AuthenticateResponse. Identity d228000a0019b615b194c5e11027942e; Path /d228000a0019b615b194c5e11027942e/tests
10-16 22:04:43.602 26369-26394/io.realm.test D/REALM_JAVA: Session[/data/data/io.realm.test/cache/junit1946518965/d228000a0019b615b194c5e11027942e/d228000a0019b615b194c5e11027942e/tests]: Access token acquired
10-16 22:04:43.603 26369-26394/io.realm.test D/REALM_JAVA: Scheduling an access_token refresh in 36739397 milliseconds
10-16 22:04:43.603 26369-26392/io.realm.test D/REALM_SYNC: Opening Realm file: /data/data/io.realm.test/cache/junit1946518965/d228000a0019b615b194c5e11027942e/d228000a0019b615b194c5e11027942e/tests
10-16 22:04:43.604 26369-26392/io.realm.test D/REALM_SYNC: Connection[1]: WebSocket::Websocket()
10-16 22:04:43.604 26369-26392/io.realm.test D/REALM_SYNC: Connection[1]: Reconnecting in 0 milliseconds
10-16 22:04:43.604 26369-26392/io.realm.test D/REALM_SYNC: Connection[1]: Session[2]: Starting session for '/data/data/io.realm.test/cache/junit1946518965/d228000a0019b615b194c5e11027942e/d228000a0019b615b194c5e11027942e/tests'
10-16 22:04:43.604 26369-26392/io.realm.test D/REALM_SYNC: Connection[1]: Session[2]: last_version_available  = 3
10-16 22:04:43.604 26369-26392/io.realm.test D/REALM_SYNC: Connection[1]: Session[2]: progress_server_version = 0
10-16 22:04:43.604 26369-26392/io.realm.test D/REALM_SYNC: Connection[1]: Session[2]: progress_client_version = 0
10-16 22:04:43.604 26369-26392/io.realm.test D/REALM_SYNC: Using already open Realm file: /data/data/io.realm.test/cache/junit1946518965/d228000a0019b615b194c5e11027942e/d228000a0019b615b194c5e11027942e/tests
10-16 22:04:43.604 26369-26392/io.realm.test D/REALM_SYNC: Connection[1]: Session[2]: Progress handler called, downloaded = 0, downloadable = 0, uploaded = 0, uploadable = 8398, progress version = 0, snapshot version = 3
10-16 22:04:43.604 26369-26392/io.realm.test D/REALM_SYNC: Connection[1]: Resolving '127.0.0.1:9080'
10-16 22:04:43.605 26369-26392/io.realm.test D/REALM_SYNC: Connection[1]: Connecting to endpoint '127.0.0.1:9080' (1/1)
10-16 22:04:43.605 26369-26392/io.realm.test I/REALM_SYNC: Connection[1]: Connected to endpoint '127.0.0.1:9080' (from '127.0.0.1:33598')
10-16 22:04:43.605 26369-26392/io.realm.test D/REALM_SYNC: Connection[1]: WebSocket::initiate_client_handshake()
10-16 22:04:43.611 26369-26392/io.realm.test D/REALM_SYNC: Connection[1]: WebSocket::handle_http_response_received()
10-16 22:04:43.612 26369-26392/io.realm.test D/REALM_SYNC: Connection[1]: Will emit a ping in 600000 milliseconds
10-16 22:04:43.612 26369-26392/io.realm.test D/REALM_SYNC: Connection[1]: Session[2]: Sending: BIND(server_path='/d228000a0019b615b194c5e11027942e/tests', signed_user_token_size=645, need_file_ident_pair=1)
10-16 22:04:43.637 26369-26392/io.realm.test D/REALM_SYNC: Connection[1]: Session[2]: Received: ALLOC(server_file_ident=1, client_file_ident=2, client_file_ident_secret=4942239746921944988)
10-16 22:04:43.637 26369-26392/io.realm.test D/REALM_SYNC: Using already open Realm file: /data/data/io.realm.test/cache/junit1946518965/d228000a0019b615b194c5e11027942e/d228000a0019b615b194c5e11027942e/tests
10-16 22:04:43.644 26369-26392/io.realm.test D/REALM_SYNC: Connection[1]: Session[2]: Sending: IDENT(server_file_ident=1, client_file_ident=2, client_file_ident_secret=4942239746921944988, scan_server_version=0, scan_client_version=0, latest_server_version=0, latest_server_session_ident=0)
10-16 22:04:43.644 26369-26392/io.realm.test D/REALM_SYNC: Using already open Realm file: /data/data/io.realm.test/cache/junit1946518965/d228000a0019b615b194c5e11027942e/d228000a0019b615b194c5e11027942e/tests
10-16 22:04:43.645 26369-26392/io.realm.test D/REALM_SYNC: Connection[1]: Session[2]: Fetching: UPLOAD changeset(client_version=2, server_version=0, changeset_size=7932, timestamp=88135483580)
10-16 22:04:43.647 26369-26392/io.realm.test D/REALM_SYNC: Connection[1]: Session[2]: Log compaction: original size = 7932, compacted size = 7932
10-16 22:04:43.647 26369-26392/io.realm.test D/REALM_SYNC: Connection[1]: Session[2]: Fetching: UPLOAD changeset(client_version=3, server_version=0, changeset_size=466, timestamp=88135483597)
10-16 22:04:43.648 26369-26392/io.realm.test D/REALM_SYNC: Connection[1]: Session[2]: Log compaction: original size = 466, compacted size = 396
10-16 22:04:43.653 26369-26392/io.realm.test D/REALM_SYNC: Sending: UPLOAD(session_ident=2, num_changesets=2, is_body_compressed=1, body_size=8369, compressed_body_size=3803)
10-16 22:04:43.654 26369-26392/io.realm.test D/REALM_SYNC: Download message compression: is_body_compressed = 0, compressed_body_size=0, uncompressed_body_size=0
10-16 22:04:43.654 26369-26392/io.realm.test D/REALM_SYNC: Connection[1]: Session[2]: Received: DOWNLOAD(scan_server_version=0, scan_client_version=0, latest_server_version=0, latest_server_session_ident=0, latest_client_version=0, downloadable_bytes=0, number_of_changesets=0)
10-16 22:04:43.654 26369-26392/io.realm.test D/REALM_SYNC: Using already open Realm file: /data/data/io.realm.test/cache/junit1946518965/d228000a0019b615b194c5e11027942e/d228000a0019b615b194c5e11027942e/tests
10-16 22:04:43.656 26369-26392/io.realm.test D/REALM_SYNC: Using already open Realm file: /data/data/io.realm.test/cache/junit1946518965/d228000a0019b615b194c5e11027942e/d228000a0019b615b194c5e11027942e/tests
10-16 22:04:43.656 26369-26392/io.realm.test D/REALM_SYNC: Connection[1]: Session[2]: Progress handler called, downloaded = 0, downloadable = 0, uploaded = 0, uploadable = 8398, progress version = 1, snapshot version = 5
10-16 22:04:43.750 26369-26392/io.realm.test D/REALM_SYNC: Download message compression: is_body_compressed = 0, compressed_body_size=0, uncompressed_body_size=0
10-16 22:04:43.750 26369-26392/io.realm.test D/REALM_SYNC: Connection[1]: Session[2]: Received: DOWNLOAD(scan_server_version=2, scan_client_version=3, latest_server_version=2, latest_server_session_ident=4324890589938338282, latest_client_version=3, downloadable_bytes=0, number_of_changesets=0)
10-16 22:04:43.750 26369-26392/io.realm.test D/REALM_SYNC: Using already open Realm file: /data/data/io.realm.test/cache/junit1946518965/d228000a0019b615b194c5e11027942e/d228000a0019b615b194c5e11027942e/tests
10-16 22:04:43.750 26369-26392/io.realm.test D/REALM_SYNC: Using already open Realm file: /data/data/io.realm.test/cache/junit1946518965/d228000a0019b615b194c5e11027942e/d228000a0019b615b194c5e11027942e/tests
10-16 22:04:43.750 26369-26392/io.realm.test D/REALM_SYNC: Connection[1]: Session[2]: Progress handler called, downloaded = 0, downloadable = 0, uploaded = 8398, uploadable = 8398, progress version = 1, snapshot version = 6
10-16 22:04:43.760 26369-26392/io.realm.test E/REALM_JAVA: createRemoteData upload 8398/8398 objects count:10
10-16 22:04:43.761 26369-26392/io.realm.test I/REALM_SYNC: Closing Realm file: /data/data/io.realm.test/cache/junit1946518965/d228000a0019b615b194c5e11027942e/d228000a0019b615b194c5e11027942e/tests
10-16 22:04:43.761 26369-26384/io.realm.test D/REALM_JAVA: last session dropped, remove network listener
10-16 22:04:43.761 26369-26392/io.realm.test D/REALM_SYNC: Connection[1]: Session[2]: Sending: UNBIND
10-16 22:04:43.761 26369-26384/io.realm.test D/REALM_JAVA: Network request (authenticate): http://127.0.0.1:9080/auth
10-16 22:04:43.873 26369-26384/io.realm.test D/REALM_JAVA: AuthenticateResponse. accessToken = null
10-16 22:04:43.873 26369-26384/io.realm.test I/REALM_JAVA: Succeeded authenticating user.
                                                           {UserId: aff5f98108801b96c0ff9a5fd66fdd76, AuthUrl: http://127.0.0.1:9080/auth}
10-16 22:04:43.878 26369-26384/io.realm.test D/REALM_JAVA: first session created add network listener
10-16 22:04:43.880 26369-26384/io.realm.test D/REALM_JNI: Callback to Java requesting token for path
10-16 22:04:43.881 26369-26397/io.realm.test D/REALM_JAVA: Network request (authenticate): http://127.0.0.1:9080/auth
10-16 22:04:43.883 26369-26384/io.realm.test D/REALM_JNI: Callback to Java requesting token for path
10-16 22:04:43.883 26369-26398/io.realm.test D/REALM_JAVA: Network request (authenticate): http://127.0.0.1:9080/auth
10-16 22:04:43.884 26369-26397/io.realm.test D/REALM_JAVA: AuthenticateResponse - Error: UNKNOWN(-1)
                                                           java.io.InterruptedIOException: thread interrupted
                                                               at okio.Timeout.throwIfReached(Timeout.java:145)
                                                               at okio.Okio$1.write(Okio.java:76)
                                                               at okio.AsyncTimeout$1.write(AsyncTimeout.java:180)
                                                               at okio.RealBufferedSink.flush(RealBufferedSink.java:216)
                                                               at okhttp3.internal.http1.Http1Codec.finishRequest(Http1Codec.java:164)
                                                               at okhttp3.internal.http.CallServerInterceptor.intercept(CallServerInterceptor.java:84)
                                                               at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:147)
                                                               at okhttp3.internal.connection.ConnectInterceptor.intercept(ConnectInterceptor.java:45)
                                                               at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:147)
                                                               at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:121)
                                                               at okhttp3.internal.cache.CacheInterceptor.intercept(CacheInterceptor.java:93)
                                                               at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:147)
                                                               at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:121)
                                                               at okhttp3.internal.http.BridgeInterceptor.intercept(BridgeInterceptor.java:93)
                                                               at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:147)
                                                               at okhttp3.internal.http.RetryAndFollowUpInterceptor.intercept(RetryAndFollowUpInterceptor.java:125)
                                                               at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:147)
                                                               at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:121)
                                                               at okhttp3.RealCall.getResponseWithInterceptorChain(RealCall.java:200)
                                                               at okhttp3.RealCall.execute(RealCall.java:77)
                                                               at io.realm.internal.network.OkHttpAuthenticationServer.authenticate(OkHttpAuthenticationServer.java:145)
                                                               at io.realm.internal.network.OkHttpAuthenticationServer.loginToRealm(OkHttpAuthenticationServer.java:70)
                                                               at io.realm.SyncSession$2.execute(SyncSession.java:560)
                                                               at io.realm.SyncSession$2.execute(SyncSession.java:556)
                                                               at io.realm.internal.network.ExponentialBackoffTask.run(ExponentialBackoffTask.java:66)
                                                               at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:428)
                                                               at java.util.concurrent.FutureTask.run(FutureTask.java:237)
                                                               at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1133)
                                                               at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:607)
                                                               at java.lang.Thread.run(Thread.java:761)
10-16 22:04:43.885 26369-26397/io.realm.test D/REALM_JAVA: Session[/data/data/io.realm.test/cache/junit1946518965/aff5f98108801b96c0ff9a5fd66fdd76/d228000a0019b615b194c5e11027942e/tests]: Failed to get access token (UNKNOWN(-1))
10-16 22:04:43.896 26369-26398/io.realm.test D/REALM_JAVA: AuthenticateResponse. Identity aff5f98108801b96c0ff9a5fd66fdd76; Path /d228000a0019b615b194c5e11027942e/tests
10-16 22:04:43.896 26369-26398/io.realm.test D/REALM_JAVA: Session[/data/data/io.realm.test/cache/junit1946518965/aff5f98108801b96c0ff9a5fd66fdd76/d228000a0019b615b194c5e11027942e/tests]: Access token acquired
10-16 22:04:43.896 26369-26398/io.realm.test D/REALM_JAVA: Scheduling an access_token refresh in 36740104 milliseconds
10-16 22:04:43.899 26369-26392/io.realm.test D/REALM_SYNC: Opening Realm file: /data/data/io.realm.test/cache/junit1946518965/aff5f98108801b96c0ff9a5fd66fdd76/d228000a0019b615b194c5e11027942e/tests
10-16 22:04:43.900 26369-26392/io.realm.test D/REALM_SYNC: Connection[2]: WebSocket::Websocket()
10-16 22:04:43.900 26369-26392/io.realm.test D/REALM_SYNC: Connection[2]: Reconnecting in 667 milliseconds
10-16 22:04:43.900 26369-26392/io.realm.test D/REALM_SYNC: Connection[2]: Session[4]: Starting session for '/data/data/io.realm.test/cache/junit1946518965/aff5f98108801b96c0ff9a5fd66fdd76/d228000a0019b615b194c5e11027942e/tests'
10-16 22:04:43.900 26369-26392/io.realm.test D/REALM_SYNC: Connection[2]: Session[4]: last_version_available  = 0
10-16 22:04:43.900 26369-26392/io.realm.test D/REALM_SYNC: Connection[2]: Session[4]: progress_server_version = 0
10-16 22:04:43.900 26369-26392/io.realm.test D/REALM_SYNC: Connection[2]: Session[4]: progress_client_version = 0
10-16 22:04:43.900 26369-26392/io.realm.test D/REALM_SYNC: Using already open Realm file: /data/data/io.realm.test/cache/junit1946518965/aff5f98108801b96c0ff9a5fd66fdd76/d228000a0019b615b194c5e11027942e/tests
10-16 22:04:43.900 26369-26392/io.realm.test D/REALM_SYNC: Connection[2]: Session[4]: Progress handler called, downloaded = 0, downloadable = 0, uploaded = 0, uploadable = 0, progress version = 0, snapshot version = 1
10-16 22:04:43.921 26369-26392/io.realm.test D/REALM_SYNC: Using already open Realm file: /data/data/io.realm.test/cache/junit1946518965/aff5f98108801b96c0ff9a5fd66fdd76/d228000a0019b615b194c5e11027942e/tests
10-16 22:04:43.921 26369-26392/io.realm.test D/REALM_SYNC: Connection[2]: Session[4]: Progress handler called, downloaded = 0, downloadable = 0, uploaded = 0, uploadable = 7932, progress version = 0, snapshot version = 2
10-16 22:04:43.928 26369-26384/io.realm.test D/REALM_JNI: Callback to Java requesting token for path
10-16 22:04:43.928 26369-26384/io.realm.test D/REALM_JAVA: Scheduling an access_token refresh in 36739072 milliseconds
10-16 22:04:43.928 26369-26392/io.realm.test D/REALM_SYNC: Opening Realm file: /data/data/io.realm.test/cache/junit1946518965/d228000a0019b615b194c5e11027942e/d228000a0019b615b194c5e11027942e/tests
10-16 22:04:43.929 26369-26392/io.realm.test D/REALM_SYNC: Connection[3]: WebSocket::Websocket()
10-16 22:04:43.929 26369-26392/io.realm.test D/REALM_SYNC: Connection[3]: Reconnecting in 0 milliseconds
10-16 22:04:43.929 26369-26392/io.realm.test D/REALM_SYNC: Connection[3]: Session[5]: Starting session for '/data/data/io.realm.test/cache/junit1946518965/d228000a0019b615b194c5e11027942e/d228000a0019b615b194c5e11027942e/tests'
10-16 22:04:43.929 26369-26392/io.realm.test D/REALM_SYNC: Connection[3]: Session[5]: last_version_available  = 6
10-16 22:04:43.929 26369-26392/io.realm.test D/REALM_SYNC: Connection[3]: Session[5]: progress_server_version = 2
10-16 22:04:43.929 26369-26392/io.realm.test D/REALM_SYNC: Connection[3]: Session[5]: progress_client_version = 3
10-16 22:04:43.929 26369-26392/io.realm.test D/REALM_SYNC: Using already open Realm file: /data/data/io.realm.test/cache/junit1946518965/d228000a0019b615b194c5e11027942e/d228000a0019b615b194c5e11027942e/tests
10-16 22:04:43.929 26369-26392/io.realm.test D/REALM_SYNC: Connection[3]: Session[5]: Progress handler called, downloaded = 0, downloadable = 0, uploaded = 8398, uploadable = 8398, progress version = 0, snapshot version = 6
10-16 22:04:43.929 26369-26392/io.realm.test D/REALM_SYNC: Connection[3]: Resolving '127.0.0.1:9080'
10-16 22:04:43.929 26369-26392/io.realm.test D/REALM_SYNC: Connection[3]: Connecting to endpoint '127.0.0.1:9080' (1/1)
10-16 22:04:43.930 26369-26392/io.realm.test I/REALM_SYNC: Connection[3]: Connected to endpoint '127.0.0.1:9080' (from '127.0.0.1:33600')
10-16 22:04:43.930 26369-26392/io.realm.test D/REALM_SYNC: Connection[3]: WebSocket::initiate_client_handshake()
10-16 22:04:43.933 26369-26392/io.realm.test D/REALM_SYNC: Connection[3]: WebSocket::handle_http_response_received()
10-16 22:04:43.933 26369-26392/io.realm.test D/REALM_SYNC: Connection[3]: Will emit a ping in 600000 milliseconds
10-16 22:04:43.933 26369-26392/io.realm.test D/REALM_SYNC: Connection[3]: Session[5]: Sending: BIND(server_path='/d228000a0019b615b194c5e11027942e/tests', signed_user_token_size=645, need_file_ident_pair=0)
10-16 22:04:43.933 26369-26392/io.realm.test D/REALM_SYNC: Connection[3]: Session[5]: Sending: IDENT(server_file_ident=1, client_file_ident=2, client_file_ident_secret=4942239746921944988, scan_server_version=2, scan_client_version=3, latest_server_version=2, latest_server_session_ident=4324890589938338282)
10-16 22:04:43.933 26369-26392/io.realm.test D/REALM_SYNC: Using already open Realm file: /data/data/io.realm.test/cache/junit1946518965/d228000a0019b615b194c5e11027942e/d228000a0019b615b194c5e11027942e/tests
10-16 22:04:43.977 26369-26392/io.realm.test D/REALM_SYNC: Download message compression: is_body_compressed = 0, compressed_body_size=0, uncompressed_body_size=0
10-16 22:04:43.977 26369-26392/io.realm.test D/REALM_SYNC: Connection[3]: Session[5]: Received: DOWNLOAD(scan_server_version=2, scan_client_version=3, latest_server_version=2, latest_server_session_ident=4324890589938338282, latest_client_version=3, downloadable_bytes=0, number_of_changesets=0)
10-16 22:04:43.977 26369-26392/io.realm.test D/REALM_SYNC: Using already open Realm file: /data/data/io.realm.test/cache/junit1946518965/d228000a0019b615b194c5e11027942e/d228000a0019b615b194c5e11027942e/tests
10-16 22:04:43.979 26369-26392/io.realm.test D/REALM_SYNC: Using already open Realm file: /data/data/io.realm.test/cache/junit1946518965/d228000a0019b615b194c5e11027942e/d228000a0019b615b194c5e11027942e/tests
10-16 22:04:43.979 26369-26392/io.realm.test D/REALM_SYNC: Connection[3]: Session[5]: Progress handler called, downloaded = 0, downloadable = 0, uploaded = 8398, uploadable = 8398, progress version = 1, snapshot version = 7
10-16 22:04:44.568 26369-26392/io.realm.test D/REALM_SYNC: Connection[2]: Resolving '127.0.0.1:9080'
10-16 22:04:44.580 26369-26392/io.realm.test D/REALM_SYNC: Connection[2]: Connecting to endpoint '127.0.0.1:9080' (1/1)
10-16 22:04:44.580 26369-26392/io.realm.test I/REALM_SYNC: Connection[2]: Connected to endpoint '127.0.0.1:9080' (from '127.0.0.1:33602')
10-16 22:04:44.580 26369-26392/io.realm.test D/REALM_SYNC: Connection[2]: WebSocket::initiate_client_handshake()
10-16 22:04:44.584 26369-26392/io.realm.test D/REALM_SYNC: Connection[2]: WebSocket::handle_http_response_received()
10-16 22:04:44.584 26369-26392/io.realm.test D/REALM_SYNC: Connection[2]: Will emit a ping in 600000 milliseconds
10-16 22:04:44.584 26369-26392/io.realm.test D/REALM_SYNC: Connection[2]: Session[4]: Sending: BIND(server_path='/d228000a0019b615b194c5e11027942e/tests', signed_user_token_size=645, need_file_ident_pair=1)
10-16 22:04:44.594 26369-26392/io.realm.test D/REALM_SYNC: Connection[2]: Session[4]: Received: ALLOC(server_file_ident=1, client_file_ident=3, client_file_ident_secret=5238695896814282186)
10-16 22:04:44.594 26369-26392/io.realm.test D/REALM_SYNC: Using already open Realm file: /data/data/io.realm.test/cache/junit1946518965/aff5f98108801b96c0ff9a5fd66fdd76/d228000a0019b615b194c5e11027942e/tests
10-16 22:04:44.603 26369-26392/io.realm.test D/REALM_SYNC: Connection[2]: Session[4]: Sending: IDENT(server_file_ident=1, client_file_ident=3, client_file_ident_secret=5238695896814282186, scan_server_version=0, scan_client_version=0, latest_server_version=0, latest_server_session_ident=0)
10-16 22:04:44.603 26369-26392/io.realm.test D/REALM_SYNC: Using already open Realm file: /data/data/io.realm.test/cache/junit1946518965/aff5f98108801b96c0ff9a5fd66fdd76/d228000a0019b615b194c5e11027942e/tests
10-16 22:04:44.603 26369-26392/io.realm.test D/REALM_SYNC: Connection[2]: Session[4]: Fetching: UPLOAD changeset(client_version=2, server_version=0, changeset_size=7932, timestamp=88135483907)
10-16 22:04:44.606 26369-26392/io.realm.test D/REALM_SYNC: Connection[2]: Session[4]: Log compaction: original size = 7932, compacted size = 7932
10-16 22:04:44.606 26369-26392/io.realm.test D/REALM_SYNC: Sending: UPLOAD(session_ident=4, num_changesets=1, is_body_compressed=1, body_size=7953, compressed_body_size=3617)
10-16 22:04:44.610 26369-26392/io.realm.test D/REALM_SYNC: Download message compression: is_body_compressed = 1, compressed_body_size=3805, uncompressed_body_size=8382
10-16 22:04:44.610 26369-26392/io.realm.test D/REALM_SYNC: Connection[2]: Session[4]: Received: DOWNLOAD(scan_server_version=2, scan_client_version=0, latest_server_version=2, latest_server_session_ident=4324890589938338282, latest_client_version=0, downloadable_bytes=8328, number_of_changesets=2)
10-16 22:04:44.610 26369-26392/io.realm.test D/REALM_SYNC: Using already open Realm file: /data/data/io.realm.test/cache/junit1946518965/aff5f98108801b96c0ff9a5fd66fdd76/d228000a0019b615b194c5e11027942e/tests
10-16 22:04:44.648 26369-26392/io.realm.test D/REALM_SYNC: Connection[2]: Session[4]: 2 remote changesets integrated, producing client version 4
10-16 22:04:44.648 26369-26392/io.realm.test D/REALM_SYNC: Using already open Realm file: /data/data/io.realm.test/cache/junit1946518965/aff5f98108801b96c0ff9a5fd66fdd76/d228000a0019b615b194c5e11027942e/tests
10-16 22:04:44.648 26369-26392/io.realm.test D/REALM_SYNC: Connection[2]: Session[4]: Progress handler called, downloaded = 8328, downloadable = 8328, uploaded = 0, uploadable = 7932, progress version = 1, snapshot version = 4
10-16 22:04:44.652 26369-26392/io.realm.test E/REALM_JAVA: downloadProgressListener_indefinitely download 8328/8328 objects count:0
10-16 22:04:44.652 26369-26392/io.realm.test D/REALM_SYNC: Download message compression: is_body_compressed = 0, compressed_body_size=0, uncompressed_body_size=0
10-16 22:04:44.652 26369-26392/io.realm.test D/REALM_SYNC: Connection[2]: Session[4]: Received: DOWNLOAD(scan_server_version=3, scan_client_version=2, latest_server_version=3, latest_server_session_ident=4324890589938338282, latest_client_version=2, downloadable_bytes=8328, number_of_changesets=0)
10-16 22:04:44.653 26369-26392/io.realm.test D/REALM_SYNC: Using already open Realm file: /data/data/io.realm.test/cache/junit1946518965/aff5f98108801b96c0ff9a5fd66fdd76/d228000a0019b615b194c5e11027942e/tests
10-16 22:04:44.654 26369-26392/io.realm.test D/REALM_SYNC: Using already open Realm file: /data/data/io.realm.test/cache/junit1946518965/aff5f98108801b96c0ff9a5fd66fdd76/d228000a0019b615b194c5e11027942e/tests
10-16 22:04:44.654 26369-26392/io.realm.test D/REALM_SYNC: Connection[2]: Session[4]: Progress handler called, downloaded = 8328, downloadable = 8328, uploaded = 7932, uploadable = 7932, progress version = 1, snapshot version = 5
10-16 22:04:44.654 26369-26392/io.realm.test D/REALM_SYNC: Download message compression: is_body_compressed = 1, compressed_body_size=2538, uncompressed_body_size=5302
10-16 22:04:44.654 26369-26392/io.realm.test D/REALM_SYNC: Connection[3]: Session[5]: Received: DOWNLOAD(scan_server_version=3, scan_client_version=3, latest_server_version=3, latest_server_session_ident=4324890589938338282, latest_client_version=3, downloadable_bytes=5274, number_of_changesets=1)
10-16 22:04:44.654 26369-26392/io.realm.test D/REALM_SYNC: Using already open Realm file: /data/data/io.realm.test/cache/junit1946518965/d228000a0019b615b194c5e11027942e/d228000a0019b615b194c5e11027942e/tests
10-16 22:04:44.657 26369-26392/io.realm.test D/REALM_SYNC: Connection[3]: Session[5]: 1 remote changeset integrated, producing client version 8
10-16 22:04:44.657 26369-26392/io.realm.test D/REALM_SYNC: Using already open Realm file: /data/data/io.realm.test/cache/junit1946518965/d228000a0019b615b194c5e11027942e/d228000a0019b615b194c5e11027942e/tests
10-16 22:04:44.657 26369-26392/io.realm.test D/REALM_SYNC: Connection[3]: Session[5]: Progress handler called, downloaded = 5274, downloadable = 5274, uploaded = 8398, uploadable = 8398, progress version = 1, snapshot version = 8
10-16 22:04:44.659 26369-26396/io.realm.test E/REALM_JAVA: createRemoteData upload 8398/8398 objects count:10
10-16 22:04:44.659 26369-26396/io.realm.test W/System.err: java.lang.AssertionError: expected:<20> but was:<10>
10-16 22:04:44.660 26369-26396/io.realm.test W/System.err:     at org.junit.Assert.fail(Assert.java:88)
10-16 22:04:44.660 26369-26396/io.realm.test W/System.err:     at org.junit.Assert.failNotEquals(Assert.java:834)
10-16 22:04:44.660 26369-26396/io.realm.test W/System.err:     at org.junit.Assert.assertEquals(Assert.java:645)
10-16 22:04:44.660 26369-26396/io.realm.test W/System.err:     at org.junit.Assert.assertEquals(Assert.java:631)
10-16 22:04:44.660 26369-26396/io.realm.test W/System.err:     at io.realm.objectserver.ProgressListenerTests$1.onChange(ProgressListenerTests.java:102)
10-16 22:04:44.660 26369-26396/io.realm.test W/System.err:     at io.realm.SyncSession.notifyProgressListener(SyncSession.java:208)
10-16 22:04:44.660 26369-26396/io.realm.test W/System.err:     at io.realm.SyncManager.notifyProgressListener(SyncManager.java:326)
10-16 22:04:44.660 26369-26396/io.realm.test W/System.err:     at io.realm.SyncSession.nativeAddProgressListener(Native Method)
10-16 22:04:44.660 26369-26396/io.realm.test W/System.err:     at io.realm.SyncSession.addProgressListener(SyncSession.java:277)
10-16 22:04:44.660 26369-26396/io.realm.test W/System.err:     at io.realm.SyncSession.addUploadProgressListener(SyncSession.java:240)
10-16 22:04:44.660 26369-26396/io.realm.test W/System.err:     at io.realm.objectserver.ProgressListenerTests.createRemoteData(ProgressListenerTests.java:92)
10-16 22:04:44.660 26369-26396/io.realm.test W/System.err:     at io.realm.objectserver.ProgressListenerTests.access$100(ProgressListenerTests.java:55)
10-16 22:04:44.660 26369-26396/io.realm.test W/System.err:     at io.realm.objectserver.ProgressListenerTests$3.run(ProgressListenerTests.java:155)
10-16 22:04:44.660 26369-26396/io.realm.test W/System.err:     at java.lang.Thread.run(Thread.java:761)
```

--- a/realm/realm-library/src/syncIntegrationTest/java/io/realm/objectserver/ProgressListenerTests.java
+++ b/realm/realm-library/src/syncIntegrationTest/java/io/realm/objectserver/ProgressListenerTests.java
@@ -23,6 +23,7 @@ import org.junit.Test;
 import org.junit.runner.RunWith;
 
 import java.net.URI;
+import java.util.Locale;
 import java.util.concurrent.CountDownLatch;
 import java.util.concurrent.atomic.AtomicBoolean;
 import java.util.concurrent.atomic.AtomicInteger;
@@ -82,17 +83,32 @@ public class ProgressListenerTests extends StandardIntegrationTest {
 
     // Create remote data for a given user.
     private URI createRemoteData(SyncUser user) {
-        SyncConfiguration config = configFactory.createSyncConfigurationBuilder(user, Constants.USER_REALM).build();
+        final SyncConfiguration config = configFactory.createSyncConfigurationBuilder(user, Constants.USER_REALM)
+                .name("remote")
+                .build();
         final Realm realm = Realm.getInstance(config);
-        writeSampleData(realm);
         final CountDownLatch changesUploaded = new CountDownLatch(1);
         final SyncSession session = SyncManager.getSession(config);
-        session.addUploadProgressListener(ProgressMode.CURRENT_CHANGES, new ProgressListener() {
+        final long beforeAdd = realm.where(AllTypes.class).count();
+        writeSampleData(realm);
+
+        session.addUploadProgressListener(ProgressMode.INDEFINITELY, new ProgressListener() {
             @Override
             public void onChange(Progress progress) {
                 if (progress.isTransferComplete()) {
-                    session.removeProgressListener(this);
-                    changesUploaded.countDown();
+                    Realm realm = Realm.getInstance(config);
+                    final long afterAdd = realm.where(AllTypes.class).count();
+                    realm.close();
+
+                    RealmLog.warn(String.format(Locale.ENGLISH,"createRemoteData upload %d/%d objects count:%d",
+                            progress.getTransferredBytes(), progress.getTransferableBytes(), afterAdd));
+                    // FIXME: Remove this after https://github.com/realm/realm-object-store/issues/581
+                    if (afterAdd == TEST_SIZE + beforeAdd) {
+                        session.removeProgressListener(this);
+                        changesUploaded.countDown();
+                    } else if (afterAdd < TEST_SIZE + beforeAdd) {
+                        fail("The added objects are more than expected.");
+                    }
                 }
             }
         });
@@ -148,50 +164,51 @@ public class ProgressListenerTests extends StandardIntegrationTest {
         worker.start();
 
         SyncUser adminUser = UserFactory.createAdminUser(Constants.AUTH_URL);
-        final SyncConfiguration adminConfig = configFactory.createSyncConfigurationBuilder(adminUser, serverUrl.toString()).build();
+        final SyncConfiguration adminConfig = configFactory.createSyncConfigurationBuilder(adminUser, serverUrl.toString())
+                .name("local")
+                .build();
         Realm adminRealm = Realm.getInstance(adminConfig);
         Realm userRealm = Realm.getInstance(configFactory.createSyncConfigurationBuilder(userWithData, Constants.USER_REALM).build()); // Keep session alive
         SyncSession session = SyncManager.getSession(adminConfig);
         session.addDownloadProgressListener(ProgressMode.INDEFINITELY, new ProgressListener() {
             @Override
             public void onChange(Progress progress) {
-                if (progress.isTransferComplete()) {
+                Realm adminRealm = Realm.getInstance(adminConfig);
+                long objectCounts = adminRealm.where(AllTypes.class).count();
+                adminRealm.close();
+                // The downloading progress listener could be triggered at the db version where only contains the meta
+                // data. So we start checking from when the first 10 objects downloaded.
+                if (objectCounts != 0 && progress.isTransferComplete()) {
+
                     switch (transferCompleted.incrementAndGet()) {
-                        case 1:
-                            // Initial trigger when registering
-                            assertTransferComplete(progress, false);
-                            break;
-                        case 2: {
+                        case 1: {
+                            assertEquals(TEST_SIZE, objectCounts);
                             assertTransferComplete(progress, true);
-                            Realm adminRealm = Realm.getInstance(adminConfig);
-                            assertEquals(TEST_SIZE, adminRealm.where(AllTypes.class).count());
-                            adminRealm.close();
                             startWorker.countDown();
                             break;
                         }
-                        case 3: {
+                        case 2: {
                             assertTransferComplete(progress, true);
-                            Realm adminRealm = Realm.getInstance(adminConfig);
-                            assertEquals(TEST_SIZE * 2, adminRealm.where(AllTypes.class).count());
-                            adminRealm.close();
+                            assertEquals(TEST_SIZE * 2, objectCounts);
                             allChangesDownloaded.countDown();
                             break;
                         }
                         default:
                             fail("Transfer complete called too many times:" + transferCompleted.get());
                     }
+                    RealmLog.warn(String.format(
+                            Locale.ENGLISH,"downloadProgressListener_indefinitely download %d/%d objects count:%d",
+                            progress.getTransferredBytes(), progress.getTransferableBytes(), objectCounts));
                 }
             }
         });
         TestHelper.awaitOrFail(allChangesDownloaded);
         adminRealm.close();
         userRealm.close();
+        // worker thread will hang if logout happens before listener triggered.
+        worker.join();
         userWithData.logout();
         adminUser.logout();
-        // FIXME sometimes the worker thread doesn't terminate
-        // causing the test thread to wait  indefinitely until it times out
-        // https://github.com/realm/realm-java/issues/5245
-        worker.join();
     }
 
     // Make sure that a ProgressListener continues to report the correct thing, even if it crashed
@@ -263,7 +280,12 @@ public class ProgressListenerTests extends StandardIntegrationTest {
         session.addUploadProgressListener(ProgressMode.INDEFINITELY, new ProgressListener() {
             @Override
             public void onChange(Progress progress) {
-                if (progress.isTransferComplete()) {
+                Realm tempRealm = Realm.getInstance(config);
+                long objectsCount = tempRealm.where(AllTypes.class).count();
+                tempRealm.close();
+                // FIXME: Remove the objectsCount checking when
+                // https://github.com/realm/realm-object-store/issues/581 gets fixed
+                if (objectsCount != 0 && progress.isTransferComplete()) {
                     switch(transferCompleted.incrementAndGet()) {
                         case 1:
                             Realm realm = Realm.getInstance(config);

